[SYCL] Optimize SYCL framework functions with -O0

maksimsab opened the pull request.

maksimsab committed:
[SYCL] Emit "sycl-framework" metadata attached to funcions/methods in sycl namespace.

maksimsab committed:
[SYCL] Add -fsycl-optimize-framework command line option in Clang Driver. Add a warning for misuse of this option.

maksimsab committed:
[SYCL] Add SYCLLowerIR/IR/PassAdaptors.h

LLVM infrastructure does not provide a way to add a
Pass Adaptor to run Function Passes on a subset of
functions in a Module.
This commit adds ModuleToSYCLFrameworkFunctionPassAdaptor
which runs passes over functions marked by "sycl-framework" metadata.

maksimsab committed:
[SYCL] Add a SYCL framework optimization pipeline to LLVM O0 default pipeline

maksimsab commented:
@elizabethandrews Could you please take a look at FE part?

maksimsab commented on llvm/include/llvm/SYCLLowerIR/IR/PassAdaptors.h:
```diff
@@ -0,0 +1,72 @@
+/// Designed to allow composition of a FunctionPass(Manager) and
+/// a ModulePassManager, by running the FunctionPass(Manager) over every
+/// "sycl-framework "function in the module.
+class ModuleToSYCLFrameworkFunctionPassAdaptor
```
I added here 2 files in `SYCLLowerIR/IR/PassAdaptors.{h,cpp}` which copies a functionality from `llvm/IR/PassManager.{h,cpp}` of a class `ModuleToFunctionPassAdaptor` but adds a modification that we run a pass on a Function which is marked with `sycl-framework` metadata.

Are we OK with such approach?

premanandrao commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -511,6 +511,11 @@ CODEGENOPT(OpaquePointers, 1, 0)
 CODEGENOPT(CtorDtorReturnThis, 1, 0)
 
+
+/// Optimize SYCL Framework functions. This are functions
```
```suggestion
/// Optimize SYCL Framework functions. These are functions
```

premanandrao commented on clang/test/CodeGenSYCL/sycl-framework-metadata.cpp:
```diff
@@ -0,0 +1,46 @@
+    void bar3() {}
+  }
+}
+
```
How about something like:

```
namespace {
  namespace sycl {
    void bar4() {}
  }
}
```

maksimsab replied:
This case is good.

There is a problem that in that case `sycl` namespace becomes ambiguous with the first namespace in the test. I will mitigate it with a separate test file.

elizabethandrews commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2990,6 +2990,8 @@ def fsycl_max_parallel_jobs_EQ : Joined<["-"], "fsycl-max-parallel-link-jobs=">,
   "on SYCL device code post-link, i.e. the generation of SPIR-V device images "
   "or AOT compilation of each device image.">;
+def fsycl_optimize_framework : Flag<["-"], "fsycl-optimize-framework">,
+  Flags<[CC1Option, CoreOption]>, HelpText<"Option used with conjunction with -O0 flag to optimize SYCL framework utility functions and left unoptimized user kernel code. (experimental)">;
```
Option used IN conjunction with -O0**

Also, should it be leave instead of left?

elizabethandrews commented on clang/test/CodeGenSYCL/sycl-framework-metadata.cpp:
```diff
@@ -0,0 +1,46 @@
+  return 123;
+}
+
+// Check that kernel is not marked with !sycl-framework metadata
```
In this case single_task is not being defined in sycl framework, so it won't have the metadata.  Can you add a test with mock headers to ensure  you are getting behavior you require?

maksimsab committed:
[SYCL] fix code review remarks

premanandrao commented on clang/test/CodeGenSYCL/sycl-framework-metadata1.cpp:
```diff
@@ -1,7 +1,9 @@
 // whose topmost namespace is sycl.
 
-// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -triple spir64-unknown-unknown -emit-llvm %s -o - | FileCheck %s
```
I think @elizabethandrews meant to also use the single_task interface from the mock header instead of the one defined below for this.

premanandrao commented on clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:
```diff
@@ -0,0 +1,27 @@
+// Check that kernel is not marked with !sycl-framework metadata
+// CHECK-NOT: @_ZTSZ4mainE6kernel() {{.*}} !sycl-framework
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(const Func &func) {
```
Suggest using the interface from the mock header here too.

maksimsab replied:
It leads to error because of ambiguity of sycl namespace and anonymous namespace.

elizabethandrews commented on clang/test/CodeGenSYCL/sycl-framework-metadata2.cpp:
```diff
@@ -0,0 +1,27 @@
+// Check that 'anonymous namespace'::sycl::* functions are not marked with
+// !sycl-framework metadata since topmost namespace is anonymous
+// instead of sycl.
+namespace {
```
Can't you add this test to sycl-framework-metadata1.cpp. Why is this second file needed?

premanandrao replied:
I think because it leads to an ambiguity with ::sycl otherwise.

maksimsab replied:
Yes, this is because of ambiguity.

maksimsab replied:
Btw, this is also the reason why I can't use a sycl::kernel_single_task mock from `Inputs/sycl.hpp`

maksimsab committed:
[SYCL] Use sycl::kernel_single_task mock in sycl-framework-metadata1.cpp

maksimsab committed:
Merge branch 'sycl' into add_new_debug_mode

mdtoguchi commented on clang/test/Driver/sycl-optimize-framework.cpp:
```diff
@@ -0,0 +1,13 @@
+
+// NO-OPT-CHECK-NOT: fsycl-optimize-framework
+
+// RUN: %clangxx -fsycl -O0 -fsycl-optimize-framework -### %s 2>&1 | FileCheck %s -check-prefix=OPT-CHECK
```
I would assume this should also work for Windows?  We should verify that the behavior is correct with `%clang_cl -Od`

maksimsab replied:
Added corresponding tests.

mdtoguchi commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -354,6 +354,8 @@ def warn_drv_sycl_target_missing : Warning<
   InGroup<SyclTarget>;
 def err_drv_no_rdc_sycl_target_missing : Error<
   "linked binaries do not contain expected '%0' target; found targets: '%1', this is not supported with '-fno-sycl-rdc'">;
+def err_drv_fsycl_wrong_optimization_options : Error<
+  "-fsycl-optimize-framework option can be used only in conjunction with -O0 option">;
```
Maybe `err_drv_fsycl_framework_requires_option` and make the option added as a parameter (so you can use the proper option name difference between Linux and MSVC style optimization disabling)

maksimsab replied:
Made the debug option a parameter.

maksimsab committed:
[SYCL] Fix RemoveFuncAttrsFromSYCLFrameworkFuncs pass

maksimsab committed:
[SYCL] Fix ModuleToSYCLFrameworkFunctionPassAdaptor

maksimsab committed:
[SYCL] Support fsycl-optimize-framework in clang_cl Driver.

maksimsab committed:
Merge branch 'sycl' into add_new_debug_mode

maksimsab committed:
[SYCL] Fix code style and test

maksimsab committed:
[SYCL] Adjust SROA usage to upstream

maksimsab committed:
[SYCL] remove test

elizabethandrews commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -1137,6 +1154,10 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
     Fn->addFnAttrs(FnAttrBuilder);
   }
 
+  if (getLangOpts().SYCLIsDevice && D && checkIfDeclaredInSYCLNamespace(D)) {
```
Shouldn't this be done only if flag is passed?

maksimsab replied:
added a check

elizabethandrews commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -2071,6 +2074,9 @@ bool CompilerInvocation::ParseCodeGenArgs(CodeGenOptions &Opts, ArgList &Args,
   if (!Opts.EmitIEEENaNCompliantInsts && !LangOptsRef.NoHonorNaNs)
     Diags.Report(diag::err_drv_amdgpu_ieee_without_no_honor_nans);
 
+  if (Args.getLastArg(options::OPT_fsycl_optimize_framework))
```
Please set using Marshalling infrastructure instead

maksimsab replied:
done

maksimsab committed:
[SYCL] add marshalling for fsycl_optimize_framework option

maksimsab committed:
[SYCL] add AddFuncAttrsFromSYCLFrameworkFuncs pass

This pass returns optnone and noinline function
attributes after performed optimizations.

premanandrao commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -836,6 +836,23 @@ static llvm::Constant *getPrologueSignature(CodeGenModule &CGM,
+/// namespace with name "sycl"
+static bool checkIfDeclaredInSYCLNamespace(const Decl *D) {
+  const DeclContext *DC = D->getDeclContext()->getEnclosingNamespaceContext();
+  const auto *ND = dyn_cast<NamespaceDecl>(DC);
```
Can this just be a cast?

maksimsab replied:
There are plenty of different Derived classes.
https://clang.llvm.org/doxygen/classclang_1_1DeclContext.html

premanandrao reviewed (commented):
One nit/question, but otherwise looks good to me.
@elizabethandrews, had done most of the FE reviews previously.  So please wait for her approval.

maksimsab committed:
[SYCL] replace fsycl-optimize-framework with
fsycl-optimize-non-user-code

maksimsab committed:
Merge branch 'sycl' into add_new_debug_mode

elizabethandrews reviewed (approved):
FE changes LGTM

maksimsab committed:
[SYCL] add Inline passes to pipeline

maksimsab committed:
[SYCL] fix code style

maksimsab commented on llvm/include/llvm/Passes/PassBuilder.h:
```diff
@@ -298,6 +302,24 @@ class PassBuilder {
+
+  /// Constructs a simplification pipeline which operates only on
+  /// "SYCL framework" functions and appends it to the given MPM.
+  /// Mostly a copy-paste of buildModuleSimplificationPipeline.
```
We could move this part into different file under some *SYCL* directory.

sarnex replied:
My main concern is about the code duplication and decoupling from the upstream unction, My understanding is we probably want to be coupled. @AlexeySachkov Do you have any opinion on this? Thanks

sarnex replied:
Yeah naturally I prefer the second option, but I doubt we can modify the original function, so maybe this is better. I want to see if Alexey has any ideas, he has more experience than me here.

AlexeySachkov replied:
In general, if we really do need some unique pipeline, we can build and maintain it (assuming that we accept the cost of such maintenance), but I doubt that there is a need for such unique pipeline.

My understanding is that by implementing this "framework optimization" feature we simply want to get some acceptable performance in `-O0` mode without sacrificing debuggability of user code much. I suppose that any simple optimization pipeline would do the trick here.

Therefore, my suggestion would be to further explore existing pipeline building functions. For example:

https://github.com/intel/llvm/blob/8c244de9e34140651bdfc2e6ffb4c221a4de1f23/llvm/include/llvm/Passes/PassBuilder.h#L161-L177

This looks like a good candidate to re-use. It is guaranteed to not have any module-level optimizations and likely will produce good enough result for us.

maksimsab replied:
Significantly simplified optimization pipeline.

maksimsab committed:
[SYCL] remove leftover and correct option's description

maksimsab committed:
Merge branch 'sycl' into add_new_debug_mode

maksimsab committed:
[SYCL] simplify optimization pipeline

Copy-Paste part is removed. Inlining part is left.

maksimsab committed:
[SYCL] remove !sycl-framework metadata

maksimsab committed:
[SYCL] fix SYCLFrameworkOptimization.h description

maksimsab committed:
[SYCL] fix code style

asudarsa commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3004,6 +3004,10 @@ def fsycl_max_parallel_jobs_EQ : Joined<["-"], "fsycl-max-parallel-link-jobs=">,
   "or AOT compilation of each device image.">;
 def : Flag<["-"], "fsycl-rdc">, Flags<[CoreOption]>, Alias<fgpu_rdc>;
 def : Flag<["-"], "fno-sycl-rdc">, Flags<[CoreOption]>, Alias<fno_gpu_rdc>;
+def fsycl_optimize_non_user_code : Flag<["-"], "fsycl-optimize-non-user-code">,
```
A better name for this option would be 'fsycl-optimize-sycl-framework-functions'. if there has already been discussion on this, please feel free to ignore my suggestion. Thanks.

maksimsab committed:
[SYCL] fix CodeGenSYCL tests

sarnex commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -352,6 +352,8 @@ def warn_drv_sycl_target_missing : Warning<
 def err_drv_no_rdc_sycl_target_missing : Error<
   "linked binaries do not contain expected '%0' target; found targets: '%1', this is not supported with '-fno-sycl-rdc'">;
+def err_drv_fsycl_wrong_optimization_options : Error<
+  "-fsycl-optimize-non-user-code option can be used only in conjunction with %0 option">;
```
```suggestion
  "-fsycl-optimize-non-user-code option can be used only in conjunction with %0">;
```
super tiny nitpick

sarnex commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3004,6 +3004,10 @@ def fsycl_max_parallel_jobs_EQ : Joined<["-"], "fsycl-max-parallel-link-jobs=">,
 def : Flag<["-"], "fno-sycl-rdc">, Flags<[CoreOption]>, Alias<fno_gpu_rdc>;
+def fsycl_optimize_non_user_code : Flag<["-"], "fsycl-optimize-non-user-code">,
+  Flags<[CC1Option, CoreOption]>, MarshallingInfoFlag<CodeGenOpts<"OptimizeSYCLFramework">>,
+  HelpText<"Option used in conjunction with -O0 flag to "
```
another super tiny nitpick: i would either make it "with the -O0 flag" or just "with -O0"

sarnex commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1985,9 +1986,23 @@ ModulePassManager PassBuilder::buildO0DefaultPipeline(OptimizationLevel Level,
 
   MPM.addPass(createModuleToFunctionPassAdaptor(AnnotationRemarksPass()));
 
+  if (PTO.OptimizeSYCLFramework) {
```
Nice!

sarnex commented on llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:
```diff
@@ -0,0 +1,24 @@
+
+PreservedAnalyses AddDebugFuncAttrs::run(Module &M, ModuleAnalysisManager &) {
+  for (Function &F : M) {
+    F.addFnAttr(Attribute::NoInline);
```
Sorry I don't understand what this pass is neded for. I thought in the front end now if you pass -O0 but enable sycl framework O0 optimization, we remove optnone/noinline from SYCL framework functions. Isn't the IR already in the desired state at initial LLVM IR creation time?

sarnex commented on llvm/include/llvm/SYCLLowerIR/SYCLFrameworkOptimization.h:
```diff
@@ -0,0 +1,31 @@
+//===----------------------------------------------------------------------===//
+/// \file
+/// This header defines utility pass for SYCL framework optimization adding
+/// back noinline and optnone function attributes since Backends might expect
```
Sorry I don't funny understand the intent of this pass. Is there a problem where the backend needs noinline/optnone on all functions if any functions use them otherwise it errors?

Also, I assume this happens after all optimization so it doesn't end up actually preventing any optimization of the functions we want to optimize right?

maksimsab replied:
I was asked to leave IR as "if it was from default debug pipeline".

Yes, it doesn't prevent optimizations of the functions we want to optimize.

AlexeySachkov replied:
> I was asked to leave IR as "if it was from default debug pipeline".

Can we get a better definition of "it was from default debug pipeline"? Can we get a motivation for that? I just don't quite understand how doing partial optimization of a module co-exists with leaving IR as "if it was from default debug pipeline". To me that extra pass looks weird and it would be great to get more technical background and motivation for its existence.

maksimsab committed:
[SYCL] add comment and fix a typo

sarnex reviewed (approved):
LGTM but might want to get final review from @AlexeySachkov

AlexeySachkov commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2277,6 +2294,12 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
 
   F->addFnAttrs(B);
 
+  if (getLangOpts().SYCLIsDevice && getCodeGenOpts().OptimizeSYCLFramework &&
```
Why can't we re-use more of existing code and simply set `ShouldAddOptNone` to `false` under this condition?

maksimsab replied:
There is a problem that in case if we set `ShouldAddOptNone` to `false` then default code path adds `noinline` attribute what is undesirable.

https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CodeGenModule.cpp#L2232-L2236

AlexeySachkov replied:
Got it, thanks

AlexeySachkov commented on clang/test/CodeGenSYCL/sycl-optimize-non-user-code-func-attrs1.cpp:
```diff
@@ -0,0 +1,50 @@
+  }
+}
+
+// #0 and #2 contain noinline and optnone func attrs, #3 does not contain them.
+// CHECK: attributes #0 = {{.*}} noinline {{.*}} optnone
```
Note that you can capture attribute IDs instead of hardcoding them and give them meaningful names to make test more self-explanatory. Related FileCheck documentation: [here](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks)

maksimsab replied:
done

AlexeySachkov commented on clang/test/Driver/sycl-optimize-framework.cpp:
```diff
@@ -0,0 +1,20 @@
+// RUN: %clang_cl -fsycl -Od -### %s 2>&1 | FileCheck %s -check-prefix=NO-OPT-WIN-CHECK
+// NO-OPT-WIN-CHECK-NOT: fsycl-optimize-non-user-code
+
+// RUN: %clang_cl -fsycl -Od -fsycl-optimize-non-user-codek -### %s 2>&1 | FileCheck %s -check-prefix=OPT-WIN-CHECK
```
```suggestion
// RUN: %clang_cl -fsycl -Od -fsycl-optimize-non-user-code -### %s 2>&1 | FileCheck %s -check-prefix=OPT-WIN-CHECK
```

AlexeySachkov commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1985,9 +1986,23 @@ ModulePassManager PassBuilder::buildO0DefaultPipeline(OptimizationLevel Level,
   MPM.addPass(createModuleToFunctionPassAdaptor(AnnotationRemarksPass()));
 
+  if (PTO.OptimizeSYCLFramework) {
+    assert(Level == OptimizationLevel::O0 &&
+           "Optimization of sycl framework should be used only with O0 "
+           "optimization level");
```
We are within `buildO0DefaultPipeline` and there is already assertion that optimization level is set to `-O0`: do we really need to duplicate it?

AlexeySachkov commented on llvm/lib/SYCLLowerIR/SYCLFrameworkOptimization.cpp:
```diff
@@ -0,0 +1,24 @@
+using namespace sycl;
+
+PreservedAnalyses AddDebugFuncAttrs::run(Module &M, ModuleAnalysisManager &) {
+  for (Function &F : M) {
+    F.addFnAttr(Attribute::NoInline);
+    F.addFnAttr(Attribute::OptimizeNone);
+  }
```
That doesn't sound right at all. So here we are modifying **all** functions in the module - why?

There are cases when `optnone` can't be added or it will produce invalid LLVM IR:

https://github.com/intel/llvm/blob/1670049aff43dbe529a15704bb66a321696a699f/clang/lib/CodeGen/CodeGenModule.cpp#L2216-L2218

Also, as I said in another comments chain, I don't fully understand the motivation and requirement to have this pass at all and I would prefer not to have and therefore not increase maintenance cost of our compiler.

Note if we decide to keep it, there are no LIT tests for it and that doesn't seem right either.

asudarsa replied:
+1.
If we do want to preserve the attributes, we might need a way to save the attributes, remove them and then retrieve them after all optimizations have been applied. Is it a good idea to move the pipeline into sycl-post-link tool where this can be done?

maksimsab replied:
sycl-post-link question.
It isn't really a good idea to do this in sycl-post-link tool because of the following reasons:
- Some optimizations require knowledge about target's traits which are provided by [TargetInfo](https://clang.llvm.org/doxygen/classclang_1_1TargetInfo.html) class instance. We don't have a setup of TargetInfo in sycl-post-link as far as I know. So we will be obliged to add a setting up of this stuff in sycl-post-link while it is already done in Clang part.
- There is some infrastructure in clang and llvm which isn't supported in sycl-post-link. For example, you can inspect what passes are being run during compilation with `-mllvm -debug-pass=Structure` (BTW, it is broken with new pass manager, but I believe it is going to be fixed). If we move optimization passes into sycl-post-link then they become invisible for clang user and debugging becomes harder.
- Speaking from architecture point of view, it looks to me that `sycl-post-link` becomes a place where we put any SYCL stuff and it becomes overcomplicated.

About `optnone` correctness issues.
Yes, you are right. `AlwaysInlineAttr` func attr should be taken into account. Motivation is requested in corresponding JIRA tracker.

maksimsab replied:
removed this pass.

maksimsab committed:
[SYCL] adjust test to error message

maksimsab committed:
[SYCL] return noinline,optnone attrs back more accurately

maksimsab committed:
[SYCL] use regex for matching function attributes in tests

maksimsab committed:
[SYCL] remove AddDebugFuncsAttrs pass

maksimsab committed:
[SYCL] remove useless assertion

maksimsab committed:
[SYCL] correct test description

maksimsab commented:
Hi @AlexeySachkov !
Do you have any additional thoughts on this one?

AlexeySachkov reviewed (approved):
LGTM

bader commented:
PR title fix suggestion: add -> Add.

asudarsa reviewed (approved):
LGTM. Thanks

maksimsab commented:
@intel/llvm-gatekeepers Can you please merge it?
